### DISCUSSION TITLE
Go agent got the log but without content

### DISCUSSION THREAD

**Discussion by akilichen** [apache/skywalking#12710] [Q&A]
Hi, there, I'm using Kratos. Now I am using `uber-go/zap` to be the application logger.

I found that the go-agent got the logs in the console without any contents. By the way, the other features are fine.

![image](https://github.com/user-attachments/assets/d7b38fbb-b29a-4667-a615-1e418c0c94c1)
![image](https://github.com/user-attachments/assets/b40909e9-c59e-4fcd-8ab2-89c23fc488a8)
![image](https://github.com/user-attachments/assets/f6765743-f012-4891-a8e6-e4d8292e2713)

Here are some messages for you guys, please help me find out my mistakes:
- Kratos:  v2.7.2
- Agent: 0.5.0
``` 
// logger for Kratos
logger := log.With(zaplog.Logger())
...
...
//  zap log instance
type ZapLogger struct {
	log  *zap.Logger
	Sync func() error
}

// Logger 配置zap日志,将zap日志库引入
func Logger() log.Logger {
	//配置zap日志库的编码器
	encoder := zapcore.EncoderConfig{
		TimeKey:        "time",
		LevelKey:       "level",
		NameKey:        "logger",
		CallerKey:      "caller",
		MessageKey:     "company-service",
		StacktraceKey:  "stack",
		EncodeTime:     zapcore.ISO8601TimeEncoder,
		LineEnding:     zapcore.DefaultLineEnding,
		EncodeLevel:    zapcore.LowercaseLevelEncoder,
		EncodeDuration: zapcore.StringDurationEncoder,
		EncodeCaller:   zapcore.ShortCallerEncoder,
	}
	return NewZapLogger(
		encoder,
		zap.NewAtomicLevelAt(zapcore.DebugLevel),
		zap.AddStacktrace(zap.NewAtomicLevelAt(zapcore.ErrorLevel)),
		zap.AddCaller(),
		zap.AddCallerSkip(2),
		zap.Development(),
	)
}

// 日志自动切割，采用 lumberjack 实现的
func getLogWriter() zapcore.WriteSyncer {
	lumberJackLogger := &lumberjack.Logger{
		Filename:   "../../company-service.log",
		MaxSize:    100,  //日志的最大大小（M）
		MaxBackups: 10,   //日志的最大保存数量
		MaxAge:     30,   //日志文件存储最大天数
		Compress:   true, //是否执行压缩
	}
	return zapcore.AddSync(lumberJackLogger)
}

// NewZapLogger return a zap logger.
func NewZapLogger(encoder zapcore.EncoderConfig, level zap.AtomicLevel, opts ...zap.Option) *ZapLogger {
	//日志切割
	writeSyncer := getLogWriter()
	//设置日志级别
	level.SetLevel(zapcore.InfoLevel)
	var core zapcore.Core
	//开发模式下打印到标准输出
	// --根据配置文件判断输出到控制台还是日志文件--
	if os.Getenv("mode") != "prod" {
		core = zapcore.NewCore(
			zapcore.NewJSONEncoder(encoder), // 编码器配置
			zapcore.NewMultiWriteSyncer(zapcore.AddSync(os.Stdout), zapcore.AddSync(writeSyncer)), // 打印到控制台和文件
			level, // 日志级别
		)
	} else {
		core = zapcore.NewCore(
			zapcore.NewConsoleEncoder(encoder),                      // 编码器配置
			zapcore.NewMultiWriteSyncer(zapcore.AddSync(os.Stdout)), // 打印到控制台
			level, // 日志级别
		)
	}
	zapLogger := zap.New(core, opts...)
	return &ZapLogger{log: zapLogger, Sync: zapLogger.Sync}
}

// Log 实现log接口
func (l *ZapLogger) Log(level log.Level, keyvals ...interface{}) error {
	if len(keyvals) == 0 || len(keyvals)%2 != 0 {
		l.log.Warn(fmt.Sprint("Key Values must appear in pairs: ", keyvals))
		return nil
	}

	var data []zap.Field
	for i := 0; i < len(keyvals); i += 2 {
		data = append(data, zap.Any(fmt.Sprint(keyvals[i]), keyvals[i+1]))
	}

	switch level {
	case log.LevelDebug:
		l.log.Debug("", data...)
	case log.LevelInfo:
		l.log.Info("", data...)
	case log.LevelWarn:
		l.log.Warn("", data...)
	case log.LevelError:
		l.log.Error("", data...)
	case log.LevelFatal:
		l.log.Fatal("", data...)
	}
	return nil
}

```
Thanks for your time~

**Comment by akilichen** [comment]
@mrproliu Now  I found that skywalking's log works, but the application's Zap log is still.
![sw自身能读到日志但读不到zaplog](https://github.com/user-attachments/assets/9c9ba650-757b-4dc8-ae7c-6a69bc9bae57)

Is there a similar case in the community？

**Reply by mrproliu** [reply]
Please provide the version of zap, currently we just tested  v1.17.0 to v1.24.0.

**Reply by akilichen** [reply]
![image](https://github.com/user-attachments/assets/8aafa6c5-022b-491a-ab23-80da0a648802)
it's v1.23.0

**Reply by mrproliu** [reply]
Ok. I took a deep look at your code, and there is no message content when you use the zap as the following code: `l.log.Debug("", data...)`. 
If you just add the key-value in the message, you can [add the keys in this configuration](https://github.com/apache/skywalking-go/blob/main/tools/go-agent/config/agent.default.yaml#L86-L87), then It could be report key-values as a label.

**Reply by akilichen** [reply]
Trying to understand your reply, I add a env param named `company-service`, but can't works.
![image](https://github.com/user-attachments/assets/3bbced53-9c82-40c2-823a-025717780705)
![image](https://github.com/user-attachments/assets/2172c47b-c900-497b-b5a6-d6488967fcd9)
![image](https://github.com/user-attachments/assets/4ef49234-4162-423f-bed5-99046011391b)
If I got something wrong?

**Reply by mrproliu** [reply]
It's not related `MessageKey` in the encoder, It's related to your messages. As I mentioned, you have wrapper the logger, right? And your message content is always empty when using `zap.log.Debug/Info/Warn/Error/Fatal`. This is the problem. 

If you want to add your key-value in the log label, you should add them. Such as, when you using the wrapped logger as following:
```
logger.Log(zap.LevelInfo, "test-key1", "test-value1", "test-key2", "test-value2")
```

If you want to show key-values in the UI, you should export them as following configuration:
```
export SW_AGENT_LOG_REPORTER_LABEL_KEYS="test-key1,test-key2"
```

**Reply by akilichen** [reply]
So If I use the  zap.log as my logger wrapper,  I can never get the logs in the “时间戳-内容”，right?
By the way, now I have export the `SW_AGENT_LOG_REPORTER_LABEL_KEYS`  configuration like `export SW_AGENT_LOG_REPORTER_LABEL_KEYS=msg`, and got the logs in label format, thanks for this:
![image](https://github.com/user-attachments/assets/527e07c2-1978-4f82-b630-501df38ea8fb)

How can I do the next and get the logs preview in the “时间戳-内容”
![image](https://github.com/user-attachments/assets/d3e8c5b2-9025-4405-b2a8-8d628a744b5a)

**Reply by mrproliu** [reply]
The single way it makes your key-value data as the content, instant of empty string.

**Reply by akilichen** [reply]
ok, I will consider this suggestion. thanks again.
